### PR TITLE
Add k8s pod info adapter (preprocessing stage) to the mixer.

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR adds an adapter that implements the `attributes` aspect to provide more detailed information on pods for consumption by other adapters within the mixer. It acts as a thin wrapper around the existing kubernetes client-go caching layer -- and is heavily "inspired" by the existing code with the istio manager for a similar purpose. The existing testdata configuration files are updated to provided example configs for testing and initial usage of the adapter.

There are portions of this PR that do not lend themselves to easy unit testing. In particular, these are the pieces that deal with the kubernetes caching layer. Interfaces were added where it made sense to achieve as much coverage as possible, but further coverage must be added with integration tests (to setup, manipulate, and tear down clusters, for instance).

The PR was manually-tested locally by pulling in a local kubeconfig file (as part of the adapter config). Test summary:

```shell
$ ./bazel-bin/cmd/server/mixs server --globalConfigFile=testdata/globalconfig.yml --serviceConfigFile=testdata/serviceconfig.yml --alsologtostderr -v=2
...
I0416 15:42:14.203948   51163 registry.go:134] Registered attributes / kubernetes
...
Starting Config API server on port 9094
Istio Mixer: version: stable-01545a2-1-gc470ad3 (build: 2017-04-16-c470ad3, status: Modified)
Starting gRPC server on port 9091
...
I0416 15:42:18.266024   51163 mutableBag.go:369] Updating attribute bag 1:
 resetting bag to empty state
  updating string attribute source.uid from '<nil>' to 'kubernetes://hello-2992298478-qv9f3.ratelimit'
  updating string attribute target.service from '<nil>' to 'svc.cluster.local'
I0416 15:42:18.266069   51163 grpcServer.go:269] Preprocessing [0]
...
I0416 15:42:18.266541   51163 <autogenerated>:4] kubernetes:getting kubeconfig from: "adapter/kubernetes/kubeconfig"
...
I0416 15:42:18.268876   51163 <autogenerated>:4] kubernetes:building new cache controller
I0416 15:42:18.268942   51163 <autogenerated>:4] kubernetes:cluster cache started
...
I0416 15:42:18.459047   51163 <autogenerated>:4] kubernetes:Add object with key: '"default/curl-2421989462-b2g2d"'
...
I0416 15:42:18.459060   51163 <autogenerated>:4] kubernetes:Add object with key: '"ratelimit/hello-2992298478-qv9f3"'
...
W0416 15:42:18.469172   51163 <autogenerated>:5] kubernetes:error getting pod for (uid: , key: ): pod '' does not exist
W0416 15:42:18.469972   51163 <autogenerated>:5] kubernetes:error getting pod for (uid: , key: ): pod '' does not exist
W0416 15:42:18.470005   51163 attrgenmgr.go:107] Generated value 'sourcePodIP' was not mapped to an attribute.
W0416 15:42:18.470014   51163 attrgenmgr.go:107] Generated value 'sourceHostIP' was not mapped to an attribute.
I0416 15:42:18.470231   51163 grpcServer.go:275] Preprocessing [0] <-- &Status{Code:0,Message:,Details:[],}
...
```

```shell
$ ./bazel-bin/cmd/client/mixc report -a target.service=svc.cluster.local --string_attributes=source.uid="kubernetes://hello-2992298478-qv9f3.ratelimit"
Report RPC returned OK
  Attribute              Type               Value
  source.labels          map[string]string  map[app:hello pod-template-hash:2992298478 version:v1]
  source.name            string             hello-2992298478-qv9f3
  source.namespace       string             ratelimit
  source.service         string             hello
  source.serviceAccount  string             default
```

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/559)

### CODE REVIEW

File: adapter/kubernetes/cache.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+func newCacheController(clientset *kubernetes.Clientset, namespace string, refreshDuration time.Duration, logger adapter.Logger) cacheController {
+	c := &controllerImpl{
+		clientset:     clientset,
+		logger:        logger,
+		mutationsChan: make(chan resourceMutation, mutationBufferSize),
+	}
+
+	c.pods = cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(opts v1.ListOptions) (runtime.Object, error) {
```
Comment (mandarjog) [inline adapter/kubernetes/cache.go]: does `namespace==""` means all namespaces?

File: adapter/kubernetes/cache.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+			WatchFunc: func(opts v1.ListOptions) (watch.Interface, error) {
+				return clientset.Pods(namespace).Watch(opts)
+			},
+		},
+		&v1.Pod{},
+		refreshDuration,
+		cache.Indexers{},
+	)
+
+	// log all events, for debug
```
Comment (mandarjog) [inline adapter/kubernetes/cache.go]: we should conditionally enable this when started in debug mode.

File: adapter/kubernetes/cache.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+// runLogger is responsible for pulling event updates off of the mutations
+// channel and logging them via the configured logger.
+func (c *controllerImpl) runLogger(stop <-chan struct{}) {
+	for {
+		select {
+		case mutation := <-c.mutationsChan:
+			err := c.log(mutation.obj, mutation.kind)
+			if err != nil {
+				c.logger.Infof("event logging failed, will retry: %v", err)
```
Comment (mandarjog) [inline adapter/kubernetes/cache.go]: glog.V check ?

File: adapter/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+	kg := &kubegen{
+		log:    env.Logger(),
+		pods:   b.pods,
+		params: *paramsProto,
+	}
+	return kg, nil
+}
+
+func newCacheFromConfig(kubeconfigPath string, refreshDuration time.Duration, log adapter.Logger) (cacheController, error) {
+	log.Infof("getting kubeconfig from: %#v", kubeconfigPath)
```
Comment (mandarjog) [inline adapter/kubernetes/kubernetes.go]: does this work while running in-cluster? Does it has the expected behaviour when you leave the kubeconfigPath blank ?

File: adapter/kubernetes/cache.go (hunk 1 above)
Comment (douglas-reid) [inline adapter/kubernetes/cache.go]: Yes. It is hard to parse out of the PR description, but in my manual testing with namespace set to "", you can see event updates across namespaces.

File: adapter/kubernetes/cache.go (hunk 3 above)
Comment (douglas-reid) [inline adapter/kubernetes/cache.go]: the logger passed into adapters (via the env) does not support leveled logging.

File: adapter/kubernetes/kubernetes.go (hunk 4 above)
Comment (douglas-reid) [inline adapter/kubernetes/kubernetes.go]: It should -- although that behavior has not been verified. See: https://sourcegraph.com/github.com/kubernetes/client-go@master/-/blob/tools/clientcmd/client_config.go#L518:1-519:1.

File: adapter/kubernetes/cache.go (hunk 1 above)
Comment (kyessenov) [inline adapter/kubernetes/cache.go]: Yes, that's correct, "" means all namespaces.

File: adapter/kubernetes/cache.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+				time.Sleep(errorDelay)
```
Comment (ayj) [inline adapter/kubernetes/cache.go]: Does it matter that this delays runLogger from stopping? Does this need to be replaced by select over `time.After(errorDelay)` and `<-stop`?

Comment (douglas-reid) [inline adapter/kubernetes/cache.go]: Sorry, didn't see this until after I merged. It is a potential concern, although I think it is very small (if the daemon takes an extra second to stop, there isn't likely to be any large impact). I can address this logic in a follow-on (i'm adding basic verbosity-level support now and will be touching this code immediately afterwards to cleanup some of mjog's concerns).
